CRM/Core/Payment/Dummy.php

```diff
@@ -223,7 +223,9 @@ public function doPayment(&$params, $component = 'contribute') {
     // does this, it needs to invoke this hook after it has done translation,
     // but before it actually starts talking to its proprietary back-end.
     if ($propertyBag->getIsRecur()) {
-      $throwAnENoticeIfNotSetAsTheseAreRequired = $propertyBag->getRecurFrequencyInterval() . $propertyBag->getRecurFrequencyUnit();
+      if (empty($params['frequency_interval']) || empty($params['frequency_unit'])) {
+        CRM_Core_Error::deprecatedWarning('contracted frequency params not passed');
```

Comment:
@eileenmcnaughton This doesn't seem right - it would fail if passed in as `recurFrequencyInterval` / `recurFrequencyUnit` which would be the case if already cast to PropertyBag or if calling code chose to use keys per PropertyBag. Since we've already cast to PropertyBag at this point it makes more sense to call the respective methods `hasRecurFrequencyInterval()` and `hasRecurFrequencyUnit()`?

Reply:
@mattwire well sadly the documentation is down - https://docs.civicrm.org/dev/en/latest/extensions/payment-processors/create/

But this check is kinda confusing cos the property bag adds complexity & also we really specifically want to check the contracted parameters - per the broken docs - not whether the property bag can interpret them.

Reply:
Ah found the docs https://docs.civicrm.org/dev/en/latest/extensions/payment-processors/paymentclass/#recurring-contribution-parameters

Reply:
@eileenmcnaughton I recently split the paymentprocessor docs into 3 pages - so see https://docs.civicrm.org/dev/en/latest/extensions/payment-processors/paymentclass/ for the bit you want.  `doPayment()` accepts either a params array or a PropertyBag - if you pass in a propertyBag the check you are adding *will* fail.

Reply:
@mattwire it shouldn't fail - because the property bag is supposed to support accessing the array values in their contracted formats - otherwise the contract isn't met  - so if that isn't working then we might need to fix it - but I think that these lines https://github.com/civicrm/civicrm-core/blob/92a32996710ed8f19ac9ff30e38c566a05bd9ac2/Civi/Payment/PropertyBag.php#L74-L75 ensure that even if the propertyBag is passed to `doPayment` then the contract is met

Reply:
My example above was also wrong.. but doing a call to `$propertyBag->has('recurFrequencyUnit')` will achieve the same result and benefit from ensuring that the params are standardized.

Reply:
> recurFrequencyUnit

Those lines ensure that `frequency_unit` is mapped to `recurFrequencyUnit` but ArrayAccess will only work on `recurFrequencyUnit`.

Reply:
@mattwire yes - cast to `propertyBag` is 100% optional - payment processor writers *may* prefer to use the `propertyBag` - it is riskier & requires more verbose code to be used safely - as we can see here - but it is appealing to some developers. 

However the calling code is expected to support the documented contact, which itself reflects long-standing expectations implemented by any number of processors. The whole point of the `propertyBag` design was to still support the array access contract  - and I'm still pretty sure the `propertyBag` has magic in it to ensure array access works on the contracted values - not just it's own values.

FWIW I think the best way to standardise parameters is to add deprecation noise to the dummy processor if the contracted properties are not passed. We don't have a mechanism to deprecate values from that set of properties - and it's hard to make the case for doing that since changing the name of various keys an array doesn't allow a big cleanup in calling code in the way our other deprecations have done

If you don't want to merge this as is I think we should comment the line out altogether in 5.69 & then address it in master without the pressure of the rc being broken

Reply:
@artfulrobot yeah - the documentation of the documented contract is likely wrong for the property bag in some parts because there was probably some guessing involved & some copy & paste. 

I was originally keen on the `PropertyBag` - I thought it was a really clever solution but when I tried to implement it the fact it errors on NULL rather than gracefully returns + the fact it doesn't supply top level support for CreditCard fields meant that 

a) Regressions happened - both in Core & in Omnipay - the one place I implemented it in core is still there but we actually loosened it up the strictness in that case, after much discussion to make it work. In Omnipay I switched back to using Array Access

b) The main promise of the `PropertyBag` was nullified - ie I couldn't use the `getter` functions because they might fail so I had to either use `has` or a property bag function that imitates arrayAccess - both of which were more laborious than using ArrayAccess

So I tried to understand the problem `PropertyBag` was supposed to solve - it's not a helper - because of the above & the problem of the long-supported values not being documented was easily solved by documenting them - without the disruption of changing our contract. You told me that the point was to be strict - which seemed to me better solved by tests. So I guess a couple of years ago I stopped seeing any real value in the property bag. It seemed disruptive & a lot of work for no clear gain. 

However, I also kinda loved the design & learnt some cool stuff from the way you implemented it which I have used elsewhere. I have thought about copying the design to Omnipay but removing the strictness so I could use the getters.

I guess on top of the technical issues /limitations I feel like `PropertyBag` has become a blocker to solving the real issues. ie what is it trying to achieve & what is the fix. And also, I am avoidant of dealing with any bugs that cross paths with PropertyBag (not only propertyBag - there are some other features / topic areas where my past experience has been such that I won't engage with issues that touch on them).

In this case the real issue was that there was a case where the required parameters were not being passed to the `doPayment` method. `PropertyBag` was escalating that to an error but no-one ever reported it. As far as I know it is not recent (I could replicate it on earlier versions) so probably only people who didn't know how to use gitlab ever saw it. If I had seen that error in any normal release I might have pretended not to see it because it involved `PropertyBag`. But because it was in 5.69 & there is a lot of change in 5.69, meaning it was likely a regression  I did dig in. Once I had 'fixed' the property bag bug I could start to see the underlying issue & test & fix that - that is now merged to 5.69 with tests. So both in time taken to understand & fix the underlying issue & the likelihood of me even trying the code in question had the opposite effect of what was intended.